presto-main/src/main/java/com/facebook/presto/memory/MemoryPool.java

```diff
@@ -70,56 +74,93 @@ public synchronized MemoryPoolInfo getInfo()
         return new MemoryPoolInfo(maxBytes, reservedBytes, reservedRevocableBytes, queryMemoryReservations, queryMemoryRevocableReservations);
     }
 
+    public void addListener(MemoryPoolListener listener)
+    {
+        listeners.add(requireNonNull(listener, "listener cannot be null"));
+    }
+
+    public void removeListener(MemoryPoolListener listener)
+    {
+        listeners.remove(requireNonNull(listener, "listener cannot be null"));
+    }
+
     /**
      * Reserves the given number of bytes. Caller should wait on the returned future, before allocating more memory.
      */
-    public synchronized ListenableFuture<?> reserve(QueryId queryId, long bytes)
+    public ListenableFuture<?> reserve(QueryId queryId, long bytes)
     {
         checkArgument(bytes >= 0, "bytes is negative");
-        if (bytes != 0) {
-            queryMemoryReservations.merge(queryId, bytes, Long::sum);
-        }
-        reservedBytes += bytes;
-        if (getFreeBytes() <= 0) {
-            if (future == null) {
-                future = SettableFuture.create();
+
+        ListenableFuture<?> result;
+        synchronized (this) {
+            if (bytes != 0) {
+                queryMemoryReservations.merge(queryId, bytes, Long::sum);
+            }
+            reservedBytes += bytes;
+            if (getFreeBytes() <= 0) {
+                if (future == null) {
+                    future = SettableFuture.create();
+                }
+                checkState(!future.isDone(), "future is already completed");
+                result = future;
+            }
+            else {
+                result = NOT_BLOCKED;
             }
-            checkState(!future.isDone(), "future is already completed");
-            return future;
         }
-        return NOT_BLOCKED;
+
+        onMemoryReserved();
```

Comment:
Instead of notifying for every reservation (possibly a huge number), why not only notify when pool is "full", and if necessary when it becomes "not_full"?